sonar-plugin/docs/how-to-write-custom-rules-in-sonar.md

```diff
@@ -0,0 +1,214 @@
+# 1 Introduction
+Welcome to the documentation for writing custom rules in Sonar for WSO2! At WSO2, we are committed to maintaining high code quality standards and continuously improving our development processes. One way we can do this is by using the SonarQube platform to analyze our codebase for potential issues. However, we may have specific coding guidelines or standards that are unique to our company and not covered by the standard checks and plugins. That's where custom rules come in. With custom rules, we can define our own rules to ensure that our code adheres to our specific standards. In this documentation, you will learn how to write custom rules for Sonar and how to integrate them into our workflow at WSO2. By following these guidelines, you can help us maintain high code quality and contribute to the continuous improvement of our development processes.
+
+# 2 Files Involved
+There are three files involved when writing custom rules for Sonar:
+1. Test file
+1. Test class
+1. Rule class
+
+## 2.1 Test file - write up some test cases
+In SonarQube, test files are used to verify that the rules you have implemented in your custom plugin are working correctly. When you write test cases for your custom rules, you should aim to cover as many scenarios as possible to ensure that the rule is working correctly and providing accurate results.
+
+In a test file for a custom rule in SonarQube, the "Compliant" and "Noncompliant" keywords are used to specify the expected result of the rule for a given code snippet. The "Compliant" keyword indicates that the rule should not produce any issues for the given code snippet, while the "Noncompliant" keyword indicates that the rule should produce at least one issue.
+
+Here is an example of how these keywords might be used in a test case which reports an issue if variable name length is less than 2 characters:
+
+```class myClass {
+
+   int a; // Noncompliant
+   boolean b; // Noncompliant
+   char c; // Noncompliant
+
+   float aa; // Compliant
+   int bb; // Compliant
+   boolean cc; // Compliant
+}
+```
+
+## 2.2 Test class - rule’s unit test
+Unit tests are used to verify the behavior of a custom rule being implemented. The CheckVerifier class, provided by the Java Analyzer rule-testing API, is used to validate the rule implementation and abstract initialization mechanisms of the analyzer. When a rule is verified, the verifier collects lines marked as noncompliant and checks that the rule raises the expected issues and only those issues.
+The following example shows how to write a test that points to your previous file with the test cases:
+
+```@Test
```

Comment:
Take `@Test` to the next line. 
correct the markdown issue in other code blocks as well.